File: backend/app/models/document_template.py
from datetime import datetime
from enum import StrEnum

from sqlalchemy import (
    Boolean,
    DateTime,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import (
    Mapped,
    mapped_column,
)

from app.db.base import Base


class DocumentTemplateType(StrEnum):
    TECHNICAL_SPECIFICATION = (
        "technical_specification"
    )


class DocumentTemplate(Base):
    __tablename__ = "document_templates"

    id: Mapped[int] = mapped_column(
        primary_key=True,
        autoincrement=True,
    )

    name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )

    template_type: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        index=True,
    )

    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    file_name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )

    storage_path: Mapped[str] = mapped_column(
        String(1000),
        nullable=False,
    )

    version: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=1,
        server_default="1",
    )

    required_variables: Mapped[
        list[str]
    ] = mapped_column(
        JSONB,
        nullable=False,
        default=list,
        server_default="[]",
    )

    is_active: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
        server_default="true",
        index=True,
    )

    archived_at: Mapped[datetime | None] = (
        mapped_column(
            DateTime(timezone=True),
            nullable=True,
            index=True,
        )
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    @property
    def is_archived(self) -> bool:
        return self.archived_at is not None

File: frontend/src/types/documentTemplate.ts
export type DocumentTemplateType =
  | "contract"
  | "technical_specification";

export interface DocumentTemplate {
  id: number;
  name: string;
  template_type: DocumentTemplateType;
  description: string | null;
  file_name: string;
  version: number;
  required_variables: string[];
  is_active: boolean;
  archived_at: string | null;
  is_archived: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateDocumentTemplatePayload {
  name: string;
  template_type: DocumentTemplateType;
  description: string | null;
  required_variables: string[];
  file: File;
}

export interface UpdateDocumentTemplatePayload {
  name?: string;
  description?: string | null;
  is_active?: boolean;
}
